ext/opcache/jit/zend_jit_ir.c

```diff
@@ -0,0 +1,16475 @@
+				}
+				if (end_inputs) {
+					ir_END_list(end_inputs);
+				}
+			} else if (smart_branch_opcode) {
+				ir_ref if_val = ir_IF(ref);
+				ir_IF_TRUE(if_val);
+				ir_END_list(true_inputs);
+				ir_IF_FALSE(if_val);
+				ir_END_list(false_inputs);
+			} else {
+				jit_set_Z_TYPE_INFO_ref(jit, jit_ZVAL_ADDR(jit, res_addr),
+					ir_ADD_U32(ir_ZEXT_U32(ref), ir_CONST_U32(IS_FALSE)));
+				ir_END_list(end_inputs);
+			}
+	    }
+	}
+
+	if (!smart_branch_opcode || exit_addr) {
+		if (end_inputs) {
+			ir_MERGE_list(end_inputs);
+		}
+	} else {
+		_zend_jit_merge_smart_branch_inputs(jit, true_label, false_label, true_inputs, false_inputs);
+	}
+
+	return 1;
+}
+
+static int zend_jit_isset_isempty_cv(zend_jit_ctx *jit, const zend_op *opline, uint32_t op1_info, zend_jit_addr op1_addr, uint8_t smart_branch_opcode, uint32_t target_label, uint32_t target_label2, const void *exit_addr)
+{
+	zend_jit_addr res_addr = RES_ADDR();
+	uint32_t true_label = -1, false_label = -1;
+	ir_ref end_inputs = IR_UNUSED, true_inputs = IR_UNUSED, false_inputs = IR_UNUSED;
+
+	// TODO: support for empty() ???
+	ZEND_ASSERT(opline->extended_value != MAY_BE_RESOURCE);
+
+	if (smart_branch_opcode && !exit_addr) {
+		if (smart_branch_opcode == ZEND_JMPZ) {
+			true_label = target_label2;
+			false_label = target_label;
+		} else if (smart_branch_opcode == ZEND_JMPNZ) {
+			true_label = target_label;
+			false_label = target_label2;
+		} else {
+			ZEND_UNREACHABLE();
+		}
+	} else {
+		res_addr = RES_ADDR();
+	}
+
+	if (op1_info & MAY_BE_REF) {
+		ir_ref ref = jit_ZVAL_ADDR(jit, op1_addr);
+		ref = jit_ZVAL_DEREF_ref(jit, ref);
+		op1_addr = ZEND_ADDR_REF_ZVAL(ref);
+	}
+
+	if (!(op1_info & (MAY_BE_UNDEF|MAY_BE_NULL))) {
+		if (exit_addr) {
+			ZEND_ASSERT(smart_branch_opcode == ZEND_JMPZ);
+		} else if (smart_branch_opcode) {
+			ir_END_list(true_inputs);
+		} else {
+			jit_set_Z_TYPE_INFO(jit, res_addr, IS_TRUE);
+			ir_END_list(end_inputs);
+		}
+	} else if (!(op1_info & (MAY_BE_ANY - MAY_BE_NULL))) {
+		if (exit_addr) {
+			ZEND_ASSERT(smart_branch_opcode == ZEND_JMPNZ);
+		} else if (smart_branch_opcode) {
+			ir_END_list(false_inputs);
+		} else {
+			jit_set_Z_TYPE_INFO(jit, res_addr, IS_FALSE);
+			ir_END_list(end_inputs);
+		}
+	} else {
+		ir_ref ref = ir_GT(jit_Z_TYPE(jit, op1_addr), ir_CONST_U8(IS_NULL));
+		if (exit_addr) {
+			if (smart_branch_opcode == ZEND_JMPNZ) {
+				ir_GUARD_NOT(ref, ir_CONST_ADDR(exit_addr));
+			} else {
+				ir_GUARD(ref, ir_CONST_ADDR(exit_addr));
+			}
+			if (end_inputs) {
+				ir_END_list(end_inputs);
+			}
+		} else if (smart_branch_opcode) {
+			ir_ref if_val = ir_IF(ref);
+			ir_IF_TRUE(if_val);
+			ir_END_list(true_inputs);
+			ir_IF_FALSE(if_val);
+			ir_END_list(false_inputs);
+		} else {
+			jit_set_Z_TYPE_INFO_ref(jit, jit_ZVAL_ADDR(jit, res_addr),
+				ir_ADD_U32(ir_ZEXT_U32(ref), ir_CONST_U32(IS_FALSE)));
+			ir_END_list(end_inputs);
+		}
+	}
+
+	if (!smart_branch_opcode || exit_addr) {
+		if (end_inputs) {
```

Comment:
Isn't this if check always false?

Reply:
> Isn't this if check always false?

No.  ``end_inputs`` might be set on line 7900 and 7909